Fix links and improve content (#2458)
TykTechnologies/tyk-docs: tyk-docs/content/getting-started/create-api.md

  - /getting-started/tutorials/create-api/
---

## What does it mean to create an API in Tyk

> You have an API with a running service and you want to use Tyk gateway in front of your API to protect and serve it.

Reviewer: ```suggestion
You have an API with a running service and you want to use Tyk Gateway in front of your API to protect and serve it.
```

> In order for Tyk Gateway to do that, you need to configure it. The minimum information you would need to provide Tyk with is the **listen path** you want your consumers to use when calling your API via Tyk and you *API URL* (your service URL) for Tyk to send the reques to. 

Reviewer: ```suggestion
In order for Tyk Gateway to do that, you need to configure it. The minimum information you would need to provide Tyk with is the **listen path** you want your consumers to use when calling your API via Tyk and your *API URL* (your service URL) for Tyk to send the request to. 
```

> In order for Tyk Gateway to do that, you need to configure it. The minimum information you would need to provide Tyk with is the **listen path** you want your consumers to use when calling your API via Tyk and your *API URL* (your service URL) for Tyk to send the request to. 

Reviewer 2: Should this be Target URL rather than API URL?

> In Tyk you will set this information and other configuration values in an object called *Tyk API definition*. Once you create this API definition in Tyk Gateway, it can start serving your consumers and forward their requests to your API. In more technical terms, Tyk gateway will listen to requests (on the listen path) and [reverse proxy](https://en.wikipedia.org/wiki/Reverse_proxy) them to your actual API (which is your backend service that listens on the target URL you set in the API definition) 

Reviewer 2: ```suggestion
This information and other configuration values are stored in an object called a *Tyk API definition*. Once you have created your Tyk API definition and deployed it in the Tyk Gateway, the Gateway can start serving your consumers, forwarding their requests to your upstream service's API.
```

> In Tyk you will set this information and a other configuration values in an object called *Tyk API definition*. Once you create this API definition in Tyk Gateway, it can start serving your consumers and forward their requests to your API serve your APIs. In more technical terms, Tyk gateway will listen to requests (on the listen path) and [reverse proxy](https://en.wikipedia.org/wiki/Reverse_proxy them to your actual API (which is your backend service that listens on the target URL you set in the API definition) 

Reviewer: ```suggestion
In Tyk you will set this information and other configuration values in an object called *Tyk API definition*. Once you create this API definition in Tyk Gateway, it can start serving your consumers and forward their requests to your API serve your APIs. In more technical terms, Tyk gateway will listen to requests (on the listen path) and [reverse proxy](https://en.wikipedia.org/wiki/Reverse_proxy) them to your actual API (which is your backend service that listens on the target URL you set in the API definition) 
```

> Please choose the tab with the product you are using for a detail guide to create Tyk API definition:

Reviewer: ```suggestion
Please choose the tab with the product you are using for a detailed guide to create Tyk API definition:
```

{{< tabs_start >}}

{{< tab_start "Cloud" >}}
<br>
<br>

> The Cloud is simply the SaaS version of the Self-Managed product, but there are a few differences. Please use the [cloud guide to create your first API]({{< ref "tyk-cloud/getting-started-tyk-cloud/first-api" >}}).

Author: @buraksekili what do you think about this?

Reviewer: Both of them look fine, and I am okay with both. Still, I'd prefer the first version: `Please make sure you follow the Cloud Get Started guide instead` because we mention differences between Self-Managed and Cloud at the beginning of the sentence. So, in my opinion, it is better to forward users to the Get Started guide instead of creating API guideline of Cloud (which looks like specific documentation of Cloud).

Author: ok i get you. I just didn't like the instruction `Please make sure you follow` , I'll ament

> The Cloud is simply the SaaS version of the Self-Managed product, but there are a few differences. Please use the [cloud guide to create your first API]({{< ref "tyk-cloud/getting-started-tyk-cloud/first-api" >}}).

Reviewer 2: ```suggestion
Tyk Cloud is the SaaS version of the Self-Managed product, though there are a few differences. Please use the Tyk Cloud [Getting Started guide]({{< ref "tyk-cloud/getting-started-tyk-cloud/first-api" >}}) to create your first API.
```

> The Cloud is simply the SaaS version of the Self-Managed product, but there are a few differences. Please follow the [Cloud guide] ({{< ref "tyk-cloud/getting-started-tyk-cloud/first-api" >}}) to get started.

Reviewer: Also, the link that `[Cloud guide]` refers to does not work when I check the preview deployment. Is the reference format correct?
```suggestion
The Cloud is simply the SaaS version of the Self-Managed product, but there are a few differences. Please follow the [Cloud guide]({{< ref "tyk-cloud/getting-started-tyk-cloud/first-api" >}}) to get started.
```

We just sent an API definition to the Tyk `/apis` endpoint. See [API definition objects]({{< ref "tyk-gateway-api/api-definition-objects" >}}) for details of all the available objects. These objects encapsulate all of the settings for an API within Tyk.

Want to learn more from one of our team of engineers?

{{< button_left href="https://tyk.io/book-a-demo" color="green" content="Book a demo" >}}

{{< tab_end >}}

{{< tab_start "Open Source" >}}
<br>
<br>
{{< note success >}}
**Note**

> **Integration with your Open API documentation**

Reviewer 2: ```suggestion
**Integration with your OpenAPI documentation**
```

> In Tyk v4.1 we introduced an integration with [Open API Specification](https://swagger.io/specification/) (OAS) which is used as a new format for *Tyk API definition*. We call it object *Tyk OAS API Definitions*. You can import your own OpenAPI document and Tyk will know to interpret it. For a detailed tutorial, on how to use it check the guide to [creating a Tyk OAS API Definitions]({{< ref "getting-started/using-oas-definitions/create-an-oas-api#tutorial-create-an-oas-api-with-the-tyk-gateway-api" >}}).

Reviewer 2: ```suggestion
In Tyk v4.1 we introduced support for APIs defined according to the [OpenAPI Specification v3.0.3](https://spec.openapis.org/oas/v3.0.3) (OAS). This introduces a standard way to describe the vendor-agnostic elements of an API (the OpenAPI Definition, stored as an OpenAPI Document); we take this and add Tyk-specific configuration options to create the *Tyk OAS API Definition*. You can import your own OpenAPI document and Tyk will use this to generate the Tyk OAS API Definition. For a detailed tutorial on using OAS with Tyk Gateway, check out our guide to [creating a Tyk OAS API Definition]({{< ref "getting-started/using-oas-definitions/create-an-oas-api#tutorial-create-an-oas-api-with-the-tyk-gateway-api" >}}).
```

> In Tyk v4.1 we introduced support for APIs defined according to the [OpenAPI Specification v3.0.3](https://spec.openapis.org/oas/v3.0.3) (OAS). This introduces a standard way to describe the vendor-agnostic elements of an API (the OpenAPI Definition, stored as an OpenAPI Document); we take this and add Tyk-specific configuration options to create the *Tyk OAS API Definition*. You can import your own OpenAPI document and Tyk will use this to generate the Tyk OAS API Definition. For a detailed tutorial on using OAS with Tyk Gateway, check out our guide to [creating a Tyk OAS API Definition]({{< ref "getting-started/using-oas-definitions/create-an-oas-api#tutorial-create-an-oas-api-with-the-tyk-gateway-api" >}}).

Reviewer 2: ```suggestion
In Tyk v4.1 we introduced support for APIs defined according to the [OpenAPI Specification v3.0.3](https://spec.openapis.org/oas/v3.0.3) (OAS).  
This introduces a standard way to describe the vendor-agnostic elements of an API (the OpenAPI Definition, stored as an OpenAPI Document); we take this and add Tyk-specific configuration options to create the *Tyk OAS API Definition*. You can import your own OpenAPI document and Tyk will use this to generate the Tyk OAS API Definition.  
For a detailed tutorial on using OAS with Tyk Gateway, check out our guide to [creating a Tyk OAS API Definition]({{< ref "getting-started/using-oas-definitions/create-an-oas-api#tutorial-create-an-oas-api-with-the-tyk-gateway-api" >}}).
```

{{< /note >}}

## Prerequisites

> Before you continue this tutorial, you need a running [Tyk OSS gateway]({{< ref "tyk-oss-gateway" >}}). Click the button for installing Tyk Gateway in your preferred way:

Reviewer 2: ```suggestion
Before you continue this tutorial, you will need a running [Tyk OSS gateway]({{< ref "tyk-oss-gateway" >}}). Click the button for instructions on how to install Tyk Gateway:
```

> Before you continue this tutorial, you need a running (Tyk OSS gateway)({{< ref "tyk-oss-gateway" >}}). Click the button for installting Tyk Gateway in your preferred way:

Reviewer: ```suggestion
Before you continue this tutorial, you need a running (Tyk OSS gateway)({{< ref "tyk-oss-gateway" >}}). Click the button for installing Tyk Gateway in your preferred way:
```

> {{< button_left href="https://tyk.io/sign-up/#oss" color="green" content="Install tyk gateway" >}}

Reviewer 2: ```suggestion
{{< button_left href="https://tyk.io/sign-up/#oss" color="green" content="Install Tyk Gateway" >}}
```

> ## API definition Creation Methods

Reviewer 2: ```suggestion
## Creating an API on Tyk Gateway
```

> There are two ways to configure Tyk Gateway with an API definitions:

Reviewer: ```suggestion
There are two ways to configure Tyk Gateway with an API definition:
```

> 1. [Create an API with the Tyk Gateway API]({{< ref "#tutorial-create-an-api-with-the-tyk-gateway-api" >}}) - Tyk Gateway offers APIs to provide various services among which are creating API definitions.

Reviewer 2: ```suggestion
1. [Create an API with the Tyk Gateway API]({{< ref "#tutorial-create-an-api-with-the-tyk-gateway-api" >}}) - Tyk Gateway has its own Gateway API which provides various services including the registering of Tyk API Definitions on the Gateway.
```

Author: `has its own Gateway API` - should be `has its own APIs` @andyo-tyk

> 1. [Create an API with the Tyk Gateway API]({{< ref "#tutorial-create-an-api-with-the-tyk-gateway-api" >}}) - Tyk gateway provides API to provided various services among which are creating API definitions).

Reviewer: The usage of `provided` seems incorrect. Please apply the suggested changes if it looks okay to you.
```suggestion
1. [Create an API with the Tyk Gateway API]({{< ref "#tutorial-create-an-api-with-the-tyk-gateway-api" >}}) - Tyk Gateway offers APIs to provide various services, among which are creating API definitions.
```

> 2. [Create an API in File-based Mode]({{< ref "#tutorial-create-an-api-in-file-based-mode" >}}) - By creating an API definiton file and load it to the gateway. This is demonstrated in the section.

Reviewer 2: ```suggestion
2. [Create an API in File-based Mode]({{< ref "#tutorial-create-an-api-in-file-based-mode" >}}) - alternatively you can create a Tyk API Definition in a file and then load it to the Gateway. This is demonstrated in the section.
```

Reviewer 2: "This is demonstrated in the section" - which section??

Author: My link disappeared. will fix t

> 2. [Create an API in File-based Mode]({{< ref "#tutorial-create-an-api-in-file-based-mode" >}}) - By creating an API definiton file and load it to the gateway. This is demonstrated the section .

Reviewer: ```suggestion
2. [Create an API in File-based Mode]({{< ref "#tutorial-create-an-api-in-file-based-mode" >}}) - By creating an API definiton file and load it to the gateway. This is demonstrated in the section.
```

## Tutorial: Create an API with the Tyk Gateway API

> See our video for adding an API to the Open Source Gateway via the Gateway API and Postman:

Reviewer 2: ```suggestion
See our video to see how easy it is to add an API to Tyk's Open Source Gateway using the Tyk Gateway API and Postman:
```

{{< youtube UWM2ZQoGhQA >}}

> In order to use the Gateway API you will need an API key for your Gateway APIs and one command to create the API and make it live.

Reviewer 2: ```suggestion
In order to use the Gateway API to create a Tyk API Definition you will need the API key for your deployment's Gateway API and then issue just one command to create the API and make it live.
```

### Step 1: Make sure you know your API secret

> Your Tyk Gateway API secret is stored in your `tyk.conf` file, the property is called `secret`, you will need to use this as a header called `x-tyk-authorization` to make calls to the Gateway API.

Reviewer 2: ```suggestion
The API key to access your Tyk Gateway API is stored in your `tyk.conf` file; the property is called `secret`. You will need to provide this value in a header called `x-tyk-authorization` when making calls to the Gateway API.
```

### Step 2: Create an API

> To create the API, lets send a definition to the `apis` endpoint, which will return the status and version of your Gateway. Change the `x-tyk-authorization` value and `curl` domain name and port to be the correct values for your environment.

Reviewer 2: ```suggestion
To create the API, let's send a Tyk API definition to the `/apis` endpoint on your Tyk Gateway. Remember to change the `x-tyk-authorization` value (API key) in the header of your API call and set the domain name and port to target your Tyk Gateway in the `curl` command.
```

```

If the command succeeds, you will see:
```json
{
  "key": "Hello-World",
  "status": "ok",
  "action": "added"
}
```

{{< note success >}}
**Note**

> A generated API ID will be added to Tyk API definition if it's not provided while creating an API with Tyk Gateway API.

Reviewer 2: ```suggestion
All APIs deployed on Tyk Gateway are given a unique `API ID`; if you don't provide one in the Tyk API Definition when creating the API, then an `API ID` will be generated automatically.
```

{{< /note >}}

**What did we just do?**

> We just sent an API definition to the Tyk `/apis` endpoint. API definitions are discussed in detail in the API section of this documentation. These objects encapsulate all of the settings for an API within Tyk Gateway.

Reviewer 2: ```suggestion
We just registered a new API on your Tyk Gateway by sending a Tyk API definition to your Gateway's `/apis` endpoint.  
Tyk API definitions encapsulate all of the settings for an API within Tyk Gateway and are discussed in detail in the [API section]({{< ref "/tyk-gateway-api/api-definition-objects" >}}) of this documentation.
```
